File: src/domain/viewpoint_trace_pattern_serialization.py
"""Serialization of ``TracePatternSet`` back to the persisted mapping form — the exact mirror
of ``viewpoint_trace_pattern_parsing`` so that domain → YAML → domain is an identity, and a
``{ref}`` is preserved rather than expanded (immutable value-expansion happens only at
evaluation, never on save)."""

from __future__ import annotations

from src.domain.viewpoint_trace_patterns import (
    Branches,
    BranchesRef,
    DerivedReachabilityLeaf,
    DiagnosticEdge,
    InlineBranches,
    LayerMembershipEndpoint,
    Leaf,
    LeafEndpoint,
    NoneLeaf,
    RegistryEndpoint,
    StoredEdge,
    TracePattern,
    TracePatternSet,
)


def trace_patterns_to_list(patterns: TracePatternSet) -> list[dict[str, object]]:
    return [_pattern_to_mapping(p) for p in patterns.patterns]


def _pattern_to_mapping(pattern: TracePattern) -> dict[str, object]:
    out: dict[str, object] = {
        "name": pattern.name,
        "kind": pattern.kind,
        "applies_to": list(pattern.applies_to),
        "branches": _branches_to_mapping(pattern.branches),
    }
    if pattern.shortcuts:
        out["shortcuts"] = [_diagnostic_edge_to_mapping(s) for s in pattern.shortcuts]
    out["leaf"] = _leaf_to_mapping(pattern.leaf)
    if pattern.diagnostic:
        out["diagnostic"] = True
    return out


def _branches_to_mapping(branches: Branches) -> dict[str, object]:
    match branches:
        case BranchesRef(ref=ref):
            return {"ref": ref}
        case InlineBranches(edges=edges):
            return {named.label: _stored_edge_to_mapping(named.edge) for named in edges}


def _stored_edge_to_mapping(edge: StoredEdge) -> dict[str, object]:
    return {
        "kind": edge.kind,
        "connection": edge.connection,
        "direction": edge.direction,
        "endpoint": {"type": edge.endpoint_type},
    }


def _diagnostic_edge_to_mapping(edge: DiagnosticEdge) -> dict[str, object]:
    return {
        "kind": edge.kind,
        "connection": edge.connection,
        "direction": edge.direction,
        "endpoint": {"type": edge.endpoint_type},
        "status": edge.status,
    }


def _leaf_to_mapping(leaf: Leaf) -> dict[str, object]:
    match leaf:
        case NoneLeaf():
            return {"kind": leaf.kind}
        case DerivedReachabilityLeaf(connection=connection, traversal=traversal, max_hops=max_hops, endpoint=endpoint):
            return {
                "kind": leaf.kind,
                "connection": connection,
                "traversal": traversal,
                "max_hops": max_hops,
                "endpoint": _leaf_endpoint_to_mapping(endpoint),
            }


def _leaf_endpoint_to_mapping(endpoint: LeafEndpoint) -> dict[str, object]:
    match endpoint:
        case RegistryEndpoint(registry=registry):
            return {"registry": registry}
        case LayerMembershipEndpoint(domain=domain, entity_class=entity_class):
            return {"domain": domain} if entity_class is None else {"domain": domain, "class": entity_class}

File: tools/gui/src/domain/viewpointTracePatternSerialization.ts
/**
 * Serialize/deserialize `TracePatternNode`s to the persisted `trace_patterns:` mapping form —
 * the exact mirror of `viewpoint_trace_pattern_serialization.py` / `_parsing.py` so a
 * round-trip through the backend loader is byte-faithful. Structural only: the loader is the
 * single validator (I-G8), so this never rejects — it emits what the loader will re-parse.
 */

import {
  type BranchesNode,
  type DiagnosticEdgeNode,
  type DiagnosticStatus,
  type EdgeDirection,
  type LeafEndpointNode,
  type LeafNode,
  type RealizerRegistry,
  type StoredEdgeNode,
  type TracePatternNode,
  mkDiagnosticEdge,
  mkStoredEdge,
  mkTracePattern,
} from './viewpointTracePattern'

const asRecord = (raw: unknown): Record<string, unknown> => (raw ?? {}) as Record<string, unknown>

/** Stringify an `unknown` that the grammar declares as a string field — never risk an object's
 * `[object Object]` default stringification (mirrors `stringOrNull` in the criteria serializer). */
const str = (v: unknown, fallback = ''): string => (typeof v === 'string' ? v : fallback)

const storedEdgeToMapping = (edge: StoredEdgeNode): Record<string, unknown> => ({
  kind: 'stored-edge',
  connection: edge.connection,
  direction: edge.direction,
  endpoint: { type: edge.endpointType },
})

const diagnosticEdgeToMapping = (edge: DiagnosticEdgeNode): Record<string, unknown> => ({
  kind: 'diagnostic-edge',
  connection: edge.connection,
  direction: edge.direction,
  endpoint: { type: edge.endpointType },
  status: edge.status,
})

const branchesToMapping = (branches: BranchesNode): Record<string, unknown> => {
  if (branches.kind === 'ref') return { ref: branches.ref }
  const out: Record<string, unknown> = {}
  for (const edge of branches.edges) out[edge.label] = storedEdgeToMapping(edge)
  return out
}

const leafEndpointToMapping = (endpoint: LeafEndpointNode): Record<string, unknown> =>
  endpoint.kind === 'registry'
    ? { registry: endpoint.registry }
    : endpoint.entityClass === null
      ? { domain: endpoint.domain }
      : { domain: endpoint.domain, class: endpoint.entityClass }

const leafToMapping = (leaf: LeafNode): Record<string, unknown> =>
  leaf.kind === 'none'
    ? { kind: 'none' }
    : {
        kind: 'derived-reachability',
        connection: leaf.connection,
        traversal: leaf.traversal,
        max_hops: leaf.maxHops,
        endpoint: leafEndpointToMapping(leaf.endpoint),
      }

export const tracePatternToMapping = (pattern: TracePatternNode): Record<string, unknown> => {
  const out: Record<string, unknown> = {
    name: pattern.name,
    kind: 'branch-complete-realization',
    applies_to: [...pattern.appliesTo],
    branches: branchesToMapping(pattern.branches),
  }
  if (pattern.shortcuts.length > 0) out.shortcuts = pattern.shortcuts.map(diagnosticEdgeToMapping)
  out.leaf = leafToMapping(pattern.leaf)
  if (pattern.diagnostic) out.diagnostic = true
  return out
}

const edgeDirectionFromRaw = (raw: unknown): EdgeDirection => (raw === 'outgoing' ? 'outgoing' : 'incoming')

const endpointTypeFromRaw = (raw: unknown): string => str(asRecord(raw).type)

const branchesFromMapping = (raw: unknown): BranchesNode => {
  const rec = asRecord(raw)
  if ('ref' in rec) return { kind: 'ref', ref: str(rec.ref) }
  const edges: StoredEdgeNode[] = Object.entries(rec).map(([label, edge]) => {
    const node = mkStoredEdge(label)
    const e = asRecord(edge)
    node.connection = str(e.connection)
    node.direction = edgeDirectionFromRaw(e.direction)
    node.endpointType = endpointTypeFromRaw(e.endpoint)
    return node
  })
  return { kind: 'inline', edges }
}

const leafEndpointFromMapping = (raw: unknown): LeafEndpointNode => {
  const rec = asRecord(raw)
  if ('registry' in rec) return { kind: 'registry', registry: str(rec.registry) as RealizerRegistry }
  return {
    kind: 'layer',
    domain: str(rec.domain),
    entityClass: rec.class != null ? str(rec.class) : null,
  }
}

const leafFromMapping = (raw: unknown): LeafNode => {
  const rec = asRecord(raw)
  if (rec.kind !== 'derived-reachability') return { kind: 'none' }
  return {
    kind: 'derived-reachability',
    connection: str(rec.connection),
    traversal: 'direct_and_derived',
    maxHops: typeof rec.max_hops === 'number' ? rec.max_hops : 4,
    endpoint: leafEndpointFromMapping(rec.endpoint),
  }
}

export const tracePatternFromMapping = (raw: unknown): TracePatternNode => {
  const rec = asRecord(raw)
  const pattern = mkTracePattern()
  pattern.name = str(rec.name)
  pattern.appliesTo = Array.isArray(rec.applies_to) ? rec.applies_to.map((t) => str(t)) : []
  pattern.branches = branchesFromMapping(rec.branches)
  pattern.shortcuts = Array.isArray(rec.shortcuts)
    ? rec.shortcuts.map((s) => {
        const e = asRecord(s)
        const node = mkDiagnosticEdge()
        node.connection = str(e.connection)
        node.direction = edgeDirectionFromRaw(e.direction)
        node.endpointType = endpointTypeFromRaw(e.endpoint)
        node.status = str(e.status, 'shortcut') as DiagnosticStatus
        return node
      })
    : []
  pattern.leaf = leafFromMapping(rec.leaf)
  pattern.diagnostic = rec.diagnostic === true
  return pattern
}
